### PR TITLE
Handle async errors

### PR DESCRIPTION
**Pull request by seddy**
Raising this as an internal PR to get some feedback before raising an upstream PR.

This essentially is all about error handling in dataloader. I've added three behaviours, defined in `get_policy`, and affect the behaviour of calling `get` or `get_many`:

* `raise_on_error` (default) - returns the value if OK, raises an error if there was a problem
* `return_nil_on_error` - previous behaviour and requested we keep this configurable by the maintainers. Returns value if OK, returns `nil` if there's an error
* `tuples` - returns `:ok`/`:error` tuples which can be handled by the developer. A minor extension on `raise_on_error` because I needed these internally, and +1'd by another contributor, so figured it would be easy to add.

### CODE REVIEW

File: lib/dataloader.ex [hunk 1]
```diff
@@ -115,27 +158,33 @@ defmodule Dataloader do
...
   end
 
+  defp do_get({:ok, val}, :raise_on_error), do: val
+  defp do_get({:ok, val}, :return_nil_on_error), do: val
+  defp do_get({:ok, val}, :tuples), do: {:ok, val}
+
+  defp do_get({:error, reason}, :raise_on_error), do: raise(Dataloader.GetError, inspect(reason))
+  defp do_get({:error, _reason}, :return_nil_on_error), do: nil
+  defp do_get({:error, reason}, :tuples), do: {:error, reason}
+
```
Comment (seddy) [inline lib/dataloader.ex:187]: FYI - pretty much all changes in this PR revolve around allowing us to do this ^^

File: lib/dataloader.ex [hunk 2]
```diff
@@ -40,6 +45,38 @@ defmodule Dataloader do
   have a different source used for each context. This provides an easy way to
   enforce data access rules within each context. See the `DataLoader.Ecto`
   moduledocs for more details
+
+  ## Options
+
+  There are two configuration options:
+
+  * `timeout` - The maximum timeout to wait for running a source, defaults to
+    the 1s more than the maximum timeout of all added sources. Set with care,
```
Comment (Adzz) [inline lib/dataloader.ex]: extra `the`

File: lib/dataloader.ex [hunk 3]
```diff
@@ -40,6 +45,38 @@ defmodule Dataloader do
...
+    error handling if required
+
+  These can be configured on a global way by configuring in your config.exs
+  file, e.g. to configure it to return nil on error:
+
+      config :dataloader,
+        get_policy: :return_nil_on_error
+
+  Alternatively, these can be set on individual `Dataloader` structs when
+  calling `new/1`
```
Comment (Adzz) [inline lib/dataloader.ex:79]: 😍 docs!

File: lib/dataloader.ex [hunk 4]
```diff
@@ -40,6 +45,38 @@ defmodule Dataloader do
...
+
+  ### `get_policy`
+
+  There are three implemented behaviours for this:
+
+  * `raise_on_error` (default)- If successful, calling `get/4` or `get_many/4`
+    will return the value. If there was an exception when trying to load any of
+    the data, it will raise that exception
+  * `return_nil_on_error` - Behaves similar to `raise_on_error` but will just
+    return `nil` instead of `raising`. It will still log errors
```
Comment (Adzz) [inline lib/dataloader.ex:67]: `It will still log error tuples`, extra `s` on error

Comment (seddy) [reply lib/dataloader.ex:67]: Nope! This is intentional, it does not log error tuples , it logs the error that actually happened. This is completely non-tupified behaviour, is there something I could do to make that clearer do you think?

File: lib/dataloader.ex (hunk 3 above)
Comment (seddy) [reply lib/dataloader.ex:79]: You know me :neckbeard:

File: lib/dataloader.ex [hunk 5]
```diff
@@ -91,12 +131,15 @@ defmodule Dataloader do
...
       sources =
-        dataloader.sources
-        |> pmap(
+        async_safely(__MODULE__, :run_tasks, [
+          dataloader.sources,
           fun,
-          tag: "Source",
-          timeout: dataloader_timeout(dataloader)
-        )
+          [timeout: dataloader_timeout(dataloader)]
```
Comment (oneKelvinSmith) [inline lib/dataloader.ex:137]: Nice.

File: lib/dataloader.ex (hunk 1 above)
Comment (oneKelvinSmith) [reply lib/dataloader.ex:187]: I approve of this message.

File: lib/dataloader.ex [hunk 6]
```diff
@@ -176,19 +237,71 @@ defmodule Dataloader do
...
+        apply(mod, fun, args)
       end)
 
     # The infinity is safe here because the internal
     # tasks all have their own timeout.
     Task.await(task, :infinity)
   end
+
+  @doc ~S"""
+  This helper function will call `fun`on all `items` asynchronously, returning
```
Comment (oneKelvinSmith) [inline lib/dataloader.ex]: Needs whitespace between "`fun`" and "on all".

File: lib/dataloader/ecto.ex [hunk 7]
```diff
@@ -230,28 +230,36 @@ if Code.ensure_loaded?(Ecto) do
...
     defimpl Dataloader.Source do
       def run(source) do
-        results =
-          source.batches
-          |> Dataloader.pmap(
-            &run_batch(&1, source),
-            timeout: source.options[:timeout] || 15_000,
-            tag: "Ecto batch"
-          )
+        results = Dataloader.async_safely(__MODULE__, :run_batches, [source])
```
Comment (oneKelvinSmith) [inline lib/dataloader/ecto.ex:233]: 👍

File: lib/dataloader/kv.ex [hunk 8]
```diff
@@ -59,19 +82,22 @@ defmodule Dataloader.KV do
...
-          fn {batch_key, ids} ->
-            {batch_key, source.load_function.(batch_key, ids)}
-          end,
-          []
-        )
+      fun = fn {batch_key, ids} ->
+        {batch_key, source.load_function.(batch_key, ids)}
+      end
+
+      results = Dataloader.async_safely(Dataloader, :run_tasks, [source.batches, fun])
```
Comment (oneKelvinSmith) [inline lib/dataloader/kv.ex:100]: 👍

**Review by oneKelvinSmith — APPROVED** [review]

**Comment by seddy** [comment]
Many thanks @Adzz @oneKelvinSmith - I'll close this, squash the commits and raise a proper one upstream :+1:

**Pull request closed without merging**